pil-analyzer/tests/types.rs

```diff
@@ -749,8 +749,38 @@ fn simple_struct() {
 fn simple_struct_type_arg() {
     let input = "
     struct Dot<T> { x: int, y: T }
+    
+    let f: int -> Dot<int> = |i| Dot{x: 0, y: i};
+    let n = 1;
+    let x = f(n);
+    
+    let m: bool = false;
+    let y = Dot{x: 0, y: m};
+    ";
+    type_check(input, &[("x", "", "Dot<int>"), ("y", "", "Dot<bool>")]);
+}
+
+#[test]
+#[should_panic = "Cannot unify types Dot<int> and Dot"]
```

Comment:
I don't think this is the right error. `Dot` is generic with no constraints on its type parameter. Why should it not unify? Or does the type system see `Dot` as a different type than `Dot<T>`? That should also not happen. It looks like a bug where we do not detect that `Dot` needs a type argument. Is that the same with enums?

Reply:
Yes, it is a bug, the test is simply to catch it while I was doing some explorations. Sorry, I should have put this PR back as a `Draft`.

The main problem is the following:
If I want the type of a struct `Dot<T>`, to be for example, `Dot<int>`, I need to use a `NamedType`. In that case, tests like

``` rust
fn simple_struct_type_arg() {
  let input = "
  struct Dot<T> { x: int, y: T }
  
  let f: int -> Dot<int> = |i| Dot{x: 0, y: i};
  let n = 1;
  let x = f(n);
  
  let m: bool = false;
  let y = Dot{x: 0, y: m};
  ”;
  type_check(input, &[((“x”, “”, “Dot<int>”), (“y”, “”, “Dot<bool>”)]);
}
```

work correctly. But then, this will fail:

``` rust
fn simple_struct() {
    let input = "
    struct Dot { x: int, y: int }
    let f: int -> Dot = |i| Dot{x: 0, y: i};
    let x = f(0);
    ";
    type_check(input, &[("x", "", "Dot")]);
}
```

becasuse the `Dot` in `int -> Dot` seems to be a `TypeVar` so the unifier will complain: `Cannot unify types Dot<> and Dot`. 

In the case of enums, both examples work correctly, but it is true that 

``` rust
#[test]
fn simple_enum_test() {
    let input = ”
    enum Option<T> {
        Some(T),
        None
    }
    let f: int -> Option = |i| Option::Some(i);
    let x = f(0);
    ”;
    type_check(input, &[((“x”, “”, “Option<int>”)]));
}
```
will fail in the same way (`Cannot unify types Option<>...`) if the function is `int -> Option` instead of `int -> Option<int>`.

So, we could emulate the behavior that enums have by returning a `TypeVar` or `NamedType` during type inference as needed (depending on whether there are type_args in the struct or not), but that might be more of a workaround than a solution. Also, if we want something like 
``` rust 
let f: int -> Option = |i| Option::Some(i);
```
to be valid (and typed as `Option<int>`), I think we should change how the signature is created.

what do you think?

Reply:
I have been doing some research on this issue. I honestly don't see it being problematic to force the user to do this (and for my taste it's less error prone) and I also adapted type inference so that the Struct cases accommodate these variants (type_args -> NamedType, type_args.empty() -> TypeVar). 

But if you consider it necessary, we can discuss it and I can go deeper into the task in a new clean PR (since enums already have the same problem). 

Otherwise, the PR can be reviewed.

Reply:
Regardless of what the cause is, I think we should either require full type arguments for function signatures or deduce the type arguments. But I don't think we want the current behaviour. Or what did you want to test in this test? If you agree, then please remove the test.

Reply:
Can you create an issue where this appears for the enum, provide some examples etc? I think I already have an idea where the bug is in the code.